### PR TITLE
Added users and organization support in grafana client

### PR DESCRIPTION
**Pull request by kbhonagiri16**
Extended https://github.com/adejoux/grafanaclient and added methods required for our api v1.0

### CODE REVIEW

**Review by oleksiishyman — COMMENTED** [review]
for me codes looks like https://github.com/adejoux/grafanaclient/blob/master/api.go
is there any point in rewriting existing lib, or we can just follow the guide like https://www.djouxtech.net/posts/creating-grafana-dashboards-with-golang/ ?

**Comment by kbhonagiri16** [comment]
Oleskii, yes it looks like that but we do not have all required structures and methods mentioned in http://petstore.swagger.io/?url=https://raw.githubusercontent.com/zaletniy/visualization-api/master/doc/visualization-api-1.0.0.yaml so I extended the code base

File: pkg/grafanaclient/test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package main
```
Comment (funrun11) [inline pkg/grafanaclient/test.go]: Please remove it.
If you need some kind of test please do it with standard test package https://golang.org/pkg/testing/ and build tags https://golang.org/pkg/go/build/#hdr-Build_Constraints, as far as such tests require Grafana instance and can't be called unit, but rather integration.

File: pkg/grafanaclient/api.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+      "log"
+      "net/http/cookiejar"
+      "time"
+      "io"
+      "bytes"
+)
+
+const timeout = 5
+
+// GrafanaError is a error structure to handle error messages in this library
```
Comment (funrun11) [inline pkg/grafanaclient/api.go]: Let us add source where we have borrowed this code.

**Comment by kbhonagiri16** [comment]
@oleksiishyman @zaletniy

**Comment by oleksiishyman** [comment]
github.com/stretchr/testify/assert is not specified in makefile

File: pkg/grafanaclient/api_test.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
+package grafanaclient
```
Comment (funrun11) [inline pkg/grafanaclient/api_test.go]: 1. Please mark it with in `integration` tag
// +build integration
2. Create `test-integration` goal in make file where you: spin up docker container with Grafana and predefined config file; run tests against container; delete container

File: pkg/grafanaclient/api_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+import "github.com/stretchr/testify/assert"
+import "testing"
+
+var default_Id = 1
```
Comment (funrun11) [inline pkg/grafanaclient/api_test.go]: Please use environment variables for this

**Review by funrun11 — CHANGES_REQUESTED** [review]

**Comment by funrun11** [comment]
@kbhonagiri16  please address the comments

File: pkg/grafanaclient/api_test.go (hunk 3 above)
Comment (kbhonagiri16) [inline pkg/grafanaclient/api_test.go]: For 2 I have to build an image with grafana with golang installed right?

**Comment by kbhonagiri16** [comment]
@oleksiishyman @zaletniy

File: Makefile [hunk 5]
```diff
@@ -4,6 +4,7 @@ GIT_SHA=$(shell git log -n 1 --pretty=format:%H)
 VERSION=1.0
 PROJECT_NAME=$(notdir $(basename $(PWD)))
 CODE_SUBDIR=pkg
+DOCKER_USERNAME=docker_username
```
Comment (funrun11) [inline Makefile]: DOCKER_USERNAME ?= docker_username

**Review by funrun11 — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (funrun11) [reply Makefile]: Please set if not set, but not overwrite

**Comment by oleksiishyman** [comment]
I think, that when changes would be ready for merging -> we need to squash all that bunch of commits into one to keep our commit history clear

File: pkg/grafanaclient/integration_test.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+var org_user = CreateUserInOrg{LoginOrEmail: "testme", Role: "Viewer"}
+
+var url = "http://0.0.0.0:3000"
+var user = "admin"
+var pass = "admin"
+
+func Test_DoLogon(t *testing.T) {
+	session := NewSession(user, pass, url)
+	err := session.DoLogon()
+	assert.Nil(t, err, "We are expecting no error and got one")
```
Comment (funrun11) [inline pkg/grafanaclient/integration_test.go]: Please log messages like:
  `assert.Nil(t, err, fmt.Sprintf("We are expecting no error and got one err: %s", err))`

It gives much better output like:

```
--- FAIL: Test_DoLogon (0.00s)
        Error Trace:    integration_test.go:38
	Error:      	Expected nil, but got: &url.Error{Op:"Post", URL:"http://127.0.0.1:3000/login", Err:(*net.OpError)(0xc4200e40f0)}
	Messages:   	We are expecting no error and got one err: Post http://127.0.0.1:3000/login: dial tcp 127.0.0.1:3000: getsockopt: connection refused

```

**Review by funrun11 — CHANGES_REQUESTED** [review]

**Comment by kbhonagiri16** [comment]
Done with changes @oleksiishyman @zaletniy

File: pkg/grafanaclient/api.go (hunk 2 above)
Comment (funrun11) [reply pkg/grafanaclient/api.go]: :heavy_check_mark:

File: pkg/grafanaclient/test.go (hunk 1 above)
Comment (funrun11) [reply pkg/grafanaclient/test.go]: :heavy_check_mark:

File: pkg/grafanaclient/api_test.go (hunk 3 above)
Comment (funrun11) [reply pkg/grafanaclient/api_test.go]: :heavy_check_mark:

File: pkg/grafanaclient/api_test.go (hunk 4 above)
Comment (funrun11) [reply pkg/grafanaclient/api_test.go]: @kbhonagiri16 todo

File: Makefile (hunk 5 above)
Comment (funrun11) [reply Makefile]: @kbhonagiri16 todo

File: pkg/grafanaclient/integration_test.go (hunk 6 above)
Comment (funrun11) [reply pkg/grafanaclient/integration_test.go]: :heavy_check_mark:

**Comment by funrun11** [comment]
@kbhonagiri16 please squash useless commits and check couple comments above

File: pkg/grafanaclient/api_test.go (hunk 4 above)
Comment (kbhonagiri16) [reply pkg/grafanaclient/api_test.go]: This is done check latest

Comment (funrun11) [reply pkg/grafanaclient/api_test.go]: https://gobyexample.com/environment-variables

Comment (kbhonagiri16) [reply pkg/grafanaclient/api_test.go]: https://github.com/zaletniy/visualization-api/pull/4/files#diff-db4f088788a270160bb9d4e0f004ecb1

File: Makefile (hunk 5 above)
Comment (kbhonagiri16) [reply Makefile]: DONE

Comment (funrun11) [reply Makefile]: ✔️

File: pkg/grafanaclient/api_test.go (hunk 4 above)
Comment (kbhonagiri16) [reply pkg/grafanaclient/api_test.go]: updated

**Comment by oleksiishyman** [comment]
I don't have any objections, the only thing that has to be resolved is commit squashing

File: pkg/grafanaclient/integration_test.go [hunk 7]
```diff
@@ -30,8 +31,18 @@ var org_list = OrgList{ID: 1, Name: "Main Org."}
...
+
+func getenv(key string) string{
+        value := os.Getenv(key)
+        if len(value) == 0 {
+                os.Setenv(key, "admin")
+                setenv_value := os.Getenv(key)
+                return setenv_value
+        }
+        return value
+}
```
Comment (funrun11) [inline pkg/grafanaclient/integration_test.go]: Please use something like this to about any default values
```
var url =  getenv("GRAFANA_URL")
var user = getenv("GRAFANA_USER")
var pass = getenv("GRAFANA_PASS")

func getenv(key string) string {
  value := os.Getenv(key)
  if len(value) == 0 {
    fmt.Printf("Environment variable %s in not defined \n", key)
    os.Exit(1)
  }
  return value
}
```

Comment (kbhonagiri16) [reply pkg/grafanaclient/integration_test.go]: Done

**Review by funrun11 — APPROVED** [review]

**Pull request merged**